addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/handler/DSMRBridgeHandler.java

```diff
@@ -0,0 +1,297 @@
+ * sent to one of the channels.
+ *
+ * @author M. Volaart - Initial contribution
+ * @author Hilbrand Bouwkamp - Refactored way messages are forwarded to meters. Removed availableMeters dependency.
+ */
+@NonNullByDefault
+public class DSMRBridgeHandler extends BaseBridgeHandler implements DSMREventListener {
+
+    /**
+     * Factor that will be multiplied with {@link #receivedTimeoutNanos} to get the timeout factor after which the
+     * device is set off line.
+     */
+    private static final int OFFLINE_TIMEOUT_FACTOR = 10;
+
+    private final Logger logger = LoggerFactory.getLogger(DSMRBridgeHandler.class);
+
+    /**
+     * Additional meter listeners to get received meter values.
+     */
+    private final List<P1TelegramListener> meterListeners = new ArrayList<>();
+
+    /**
+     * Serial Port Manager.
+     */
+    private final SerialPortManager serialPortManager = new SerialPortManager() {
+    };
+
+    /**
+     * The dsmrDevice managing the connection and handling telegrams.
+     */
+    private @Nullable DSMRDevice dsmrDevice;
+
+    /**
+     * Long running process that controls the DSMR device connection.
+     */
+    private @Nullable DSMRDeviceRunnable dsmrDeviceRunnable;
+    /**
+     * Thread for {@link DSMRDeviceRunnable}. A thread is used because the {@link DSMRDeviceRunnable} is a blocking
+     * process that runs as long as the thing is not disposed.
+     */
+    private @Nullable Thread dsmrDeviceThread;
+    /**
+     * Watchdog to check if messages received and restart if necessary.
+     */
+    private @Nullable ScheduledFuture<?> watchdog;
+    /**
+     * Number of nanoseconds after which a timeout is triggered when no messages received.
+     */
+    private long receivedTimeoutNanos;
+
+    /**
+     * Timestamp in nanoseconds of last P1 telegram received
+     */
+    private long telegramReceivedTimeNanos;
+
+    /**
+     * Constructor
+     *
+     * @param bridge the Bridge ThingType
+     */
+    public DSMRBridgeHandler(Bridge bridge) {
+        super(bridge);
+    }
+
+    /**
+     * The {@link DSMRBridgeHandler} does not support handling commands.
+     *
+     * @param channelUID the {@link ChannelUID} of the channel to which the command was sent
+     * @param command the {@link Command}
+     */
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        // DSMRBridgeHandler does not support commands
+    }
+
+    /**
+     * Initializes this {@link DSMRBridgeHandler}.
+     *
+     * This method will get the corresponding configuration and initialize and start the corresponding
+     * {@link DSMRDevice}.
+     */
+    @Override
+    public void initialize() {
+        DSMRDeviceConfiguration deviceConfig = getConfigAs(DSMRDeviceConfiguration.class);
+
+        logger.trace("Using configuration {}", deviceConfig);
+        updateStatus(ThingStatus.UNKNOWN);
+        receivedTimeoutNanos = TimeUnit.SECONDS.toNanos(deviceConfig.receivedTimeout);
```

Comment:
Why do you / would you calculate on a nano scale if the input is seconds?

Reply:
See previous comment